README.md
# What is `judgyprophet`?

`judgyprophet` is a Bayesian forecasting algorithm based on Prophet, that enables forecasting while using information known by the business about future events. The aim is to enable users to perform forecasting with *judgmental adjustment*, in a way that is mathematically as sound as possible.

Some events will have a big effect on your timeseries. Some of which you are aware of ahead of time. For example:

* An existing product entering a new market.
* A price change to a product.

These events will typically cause a large change in your timeseries of e.g. product sales, which a standard statistical forecast will consistently underestimate.

The business will often have good estimates (or at least better than your statistical forecast) about how these events will affect your timeseries. But this is difficult to encode into your statistical forecasting algorithm. One option is to use a regressor, but this typically works poorly. This is because you have no data on the event before it occurs, and the statistical forecast does not know how to balance the information in your regressor and trend after the event occurs (which can lead to erratic behaviour).

`judgyprophet` solves this problem by encoding the business estimate of how the event will affect the forecast (the judgmental adjustment) as a Bayesian informative prior.

Before the event occurs, this business information is used to reflect the forecast of what will happen post-event e.g. the estimated uplift in product sales once the event has happened. After the event occurs, we update what the business *thinks* will happen, with what we *see* happening in the actuals. This is done using standard Bayesian updating.


# Installation

### 1. install `judgyprophet` python package using `pip`

```
pip install judgyprophet
```

### 2. compile the `STAN` model

`judgyprophet` depends on `STAN`, whose models have to be compiled before running.

So to use `judgyprophet`, you have to compile the model. Do this in the shell using

```
python -c "from judgyprophet import JudgyProphet; JudgyProphet().compile()"
```

or in python using
```python
from judgyprophet import JudgyProphet

JudgyProphet().compile()
```

This will take a while. But you only have to run this once, after the initial install.

# Documentation

Full documentation is available on our Github Pages site [here](https://astrazeneca.github.io/judgyprophet/).

Scroll down for a quickstart tutorial.

A runnable jupyter notebook version of the quickstart tutorial is available [here](https://github.com/AstraZeneca/judgyprophet/blob/main/tutorials/Quickstart.ipynb)

# Roadmap

Some things on our roadmap:

* Currently `judgyprophet` `STAN` file is only tested on Unix-based Linux or Mac machines. We aim to fully test Windows machines ASAP.
* Option to run full MCMC, rather than just L-BFGS.
* Prediction intervals
* Regressors/holidays


# Quickstart Tutorial

Imagine your business currently operates in the US, but is launching its product in Europe. As a result it anticipates a sharp uptake in sales (which it has an estimate of). As your forecasting team, they come to you and ask you to account for this.

Let's look at how we might do this using `judgyprophet` with some example data, where we know what happened. First let's plot this:


```python
from judgyprophet.tutorials.resources import get_trend_event

example_data = get_trend_event()
p = example_data.plot.line()
```



![png](docs/output_1_0.png)



We can see that product sales increased sharply from about September 2020. Suppose it was a launch in a new market, and that the business had an initial estimate of the impact in May 2020. The business expected the slope increase to be 6.

Let's use `judgyprophet` to forecast this series from May 2020. We do this by encoding the initial business estimate as a _trend event_.


```python
from judgyprophet import JudgyProphet
import pandas as pd
import seaborn as sns

# Create the expected trend events by consulting with the business
trend_events = [
    {'name': "New market entry", 'index': '2020-09-01', 'm0': 6}
]


# Cutoff the data to May 2020
data_may2020 = example_data.loc[:"2020-05-01"]

# Make the forecast with the business estimated level event
# We have no level events, so just provide the empty list.
jp = JudgyProphet()
# Because the event is beyond the actuals, judgyprophet throws a warning.
#    This is just because the Bayesian model at the event has no actuals to learn from.
#    The event is still used in predictions.
jp.fit(
    data=data_may2020,
    level_events=[],
    trend_events=trend_events,
    # Set random seed for reproducibility
    seed=13
)
predictions = jp.predict(horizon=12)
```

    INFO:judgyprophet.judgyprophet:Rescaling onto 0-mean, 1-sd.
    WARNING:judgyprophet.judgyprophet:Post-event data for trend event New market entry less than 0 points. Event deactivated in model. Event index: 2020-09-01, training data end index: 2019-06-01 00:00:00
    WARNING:judgyprophet.utils:No active trend or level events (i.e. no event indexes overlap with data). The model will just fit a base trend to the data.


    Initial log joint probability = -3.4521
        Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes
           3      -2.92768      0.054987   8.11433e-14           1           1        7
    Optimization terminated normally:
      Convergence detected: gradient norm is below tolerance


Because we are in May 2020, the forecasting algorithm has nothing to use for learning; so just uses the business estimate. Let's plot the result:


```python
from judgyprophet.tutorials.resources import plot_forecast

plot_forecast(
    actuals=example_data,
    predictions=predictions,
    cutoff="2020-05-01",
    events=trend_events
)
```

    INFO:prophet:Disabling yearly seasonality. Run prophet with yearly_seasonality=True to override this.
    INFO:prophet:Disabling weekly seasonality. Run prophet with weekly_seasonality=True to override this.
    INFO:prophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.



    Initial log joint probability = -17.0121
    Iteration  1. Log joint probability =    10.4753. Improved by 27.4875.
    Iteration  2. Log joint probability =    12.7533. Improved by 2.27796.
    Iteration  3. Log joint probability =    25.4696. Improved by 12.7163.
    Iteration  4. Log joint probability =     26.707. Improved by 1.2374.
    Iteration  5. Log joint probability =    26.7075. Improved by 0.000514342.
    Iteration  6. Log joint probability =    26.7104. Improved by 0.00296558.
    Iteration  7. Log joint probability =    26.7122. Improved by 0.00171322.
    Iteration  8. Log joint probability =    26.7157. Improved by 0.00351772.
    Iteration  9. Log joint probability =    26.7159. Improved by 0.000208268.
    Iteration 10. Log joint probability =    26.7159. Improved by 6.64977e-05.
    Iteration 11. Log joint probability =     26.716. Improved by 6.89899e-05.
    Iteration 12. Log joint probability =     26.716. Improved by 3.06578e-05.
    Iteration 13. Log joint probability =     26.716. Improved by 8.91492e-07.
    Iteration 14. Log joint probability =     26.716. Improved by 8.71052e-09.




![png](docs/output_5_2.png)



We can see `judgyprophet` is accounting for the increased trend, but the business slightly overestimated the increase in sales due to the product launch.

Let's fast forward to January 2021, the business want to reforecast based on their estimate, and what they've seen so far for the product launch. This is where `judgyprophet` comes into its own.

Once actuals are observed after the event has taken place, `judgyprophet` updates its estimate of what the event impact is. Let's look at this in action:


```python
# Cutoff the data to January 2021
data_jan2021 = example_data.loc[:"2021-01-01"]

# Reforecast using the new actuals, not we are at Jan 2021
jp = JudgyProphet()
jp.fit(
    data=data_jan2021,
    level_events=[],
    trend_events=trend_events,
    # Set random seed for reproducibility
    seed=13
)
predictions = jp.predict(horizon=12)
```

    INFO:judgyprophet.judgyprophet:Rescaling onto 0-mean, 1-sd.
    INFO:judgyprophet.judgyprophet:Adding trend event New market entry to model. Event index: 2020-09-01, training data start index: 2019-06-01 00:00:00, training data end index: 2021-01-01 00:00:00. Initial gradient: 6. Damping: None.


    Initial log joint probability = -309.562
        Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes
          10      -1.64341   2.10244e-05   3.61281e-06           1           1       15
    Optimization terminated normally:
      Convergence detected: relative gradient magnitude is below tolerance


Now let's plot the results:


```python
plot_forecast(actuals=example_data, predictions=predictions, cutoff="2021-01-01", events=trend_events)
```

    INFO:prophet:Disabling yearly seasonality. Run prophet with yearly_seasonality=True to override this.
    INFO:prophet:Disabling weekly seasonality. Run prophet with weekly_seasonality=True to override this.
    INFO:prophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.



    Initial log joint probability = -24.5881
    Iteration  1. Log joint probability =   -1.06803. Improved by 23.5201.
    Iteration  2. Log joint probability =    11.6215. Improved by 12.6895.
    Iteration  3. Log joint probability =    36.5271. Improved by 24.9056.
    Iteration  4. Log joint probability =    37.3776. Improved by 0.850488.
    Iteration  5. Log joint probability =    37.6489. Improved by 0.271259.
    Iteration  6. Log joint probability =    37.6547. Improved by 0.00580657.
    Iteration  7. Log joint probability =    37.7831. Improved by 0.128419.
    Iteration  8. Log joint probability =    37.7884. Improved by 0.00527858.
    Iteration  9. Log joint probability =     37.789. Improved by 0.000612124.
    Iteration 10. Log joint probability =    37.7891. Improved by 9.93823e-05.
    Iteration 11. Log joint probability =    37.7902. Improved by 0.00112416.
    Iteration 12. Log joint probability =    37.7902. Improved by 3.17397e-06.
    Iteration 13. Log joint probability =    37.7902. Improved by 1.59404e-05.
    Iteration 14. Log joint probability =    37.7902. Improved by 5.06854e-07.
    Iteration 15. Log joint probability =    37.7902. Improved by 6.87792e-07.
    Iteration 16. Log joint probability =    37.7902. Improved by 4.82761e-08.
    Iteration 17. Log joint probability =    37.7902. Improved by 2.50385e-07.
    Iteration 18. Log joint probability =    37.7902. Improved by 6.60322e-09.




![png](docs/output_9_2.png)



In this case, once `judgyprophet` observes the data post-event, the Bayesian updating starts to realise the business estimate is a bit large, so it reduces it.

This was a simple example to demonstrate `judgyprophet`. You can add many trend events into a single forecasting horizon, add damping. You can also add level events – changes in the forecasting level; and seasonality see our other tutorials for details about this.


NEWS.md
# Changes to `judgyprophet` by version

## `judgyprophet 0.1.0`

* Initial implementation of judgyprophet STAN code and python wrapper.
* Added level events, trend events, damped trend events and additive/multiplicative seasonality.
* `fit` and `predict` methods.


tests/conftest.py
import warnings

import pytest

from judgyprophet import JudgyProphet, utils


def pytest_configure(config):
    """
    Ensure STAN model is compiled before tests are run.
    """
    jp = JudgyProphet()
    if jp.stan_model is None:
        utils.log_and_warn(
            (
                "Model has not been compiled yet."
                " Attempting to compile STAN code now..."
            )
        )
        jp.compile()
        # Check compile worked properly
        if jp.stan_model is None:
            utils.log_and_raise(
                (
                    "STAN model failed to compile."
                    " Please check your pystan installation is working correctly."
                    " Also check your operating system has a valid C++ compiler installed."
                ),
                ValueError,
            )


tests/test_indexing.py
"""
Test expert prophet with different indexing
"""
import numpy as np
import pandas as pd
import pytest

from judgyprophet import JudgyProphet


@pytest.fixture
def create_int_index_event_data():
    """
    Data with int index
    """
    np.random.seed(13)
    # Declare constants
    event_date = 30
    level_events = [
        {"name": "L1", "index": event_date, "c0": 5.0},
    ]
    bias = 4.0
    obs_scale = 1.0
    ds = np.arange(60)
    T = len(ds)
    y = bias + np.random.normal(scale=obs_scale, size=T)
    for level_i in level_events:
        delta_i = ds >= event_date
        y += delta_i * level_i["c0"]
    return {
        "data": pd.Series(y, index=ds),
        "level_events": level_events,
        "trend_events": [],
    }


def test_int_index(create_int_index_event_data):
    jp = JudgyProphet()
    jp.fit(**create_int_index_event_data)
    model_output = jp.get_model_output()
    assert np.isclose(model_output["level_events"][0]["c"], 5.0, atol=1e-1)
    all_unspecified_params = [
        param
        for cp in model_output["changepoint_events"]
        for param in [cp["cp_c"], cp["cp_m"]]
    ]
    assert np.allclose(all_unspecified_params, 0, atol=1e-1)
    assert np.isclose(model_output["base"]["beta_0"], 4.0, atol=1.0)
    predictions = jp.predict(horizon=5)
    # Fetch expected
    ds = np.arange(65)
    delta_i = ds >= 30
    expected = 4.0 + delta_i * 5.0
    assert np.allclose(predictions["forecast"], expected, atol=1.0)


def test_float_index(create_int_index_event_data, caplog):
    """
    Check a float index results in an error
    """
    float_index_data = create_int_index_event_data.copy()
    float_index_data["data"].index += 0.1
    jp = JudgyProphet()
    with pytest.raises(ValueError):
        jp.fit(**float_index_data)
    assert "Arg 'data' must have index of type" in caplog.text


tests/test_level_event.py
"""
Test expert prophet with level events
"""
import warnings

import numpy as np
import pandas as pd
import pytest

from judgyprophet import JudgyProphet


@pytest.fixture
def create_single_event_data():
    np.random.seed(13)
    # Declare constants
    event_date = "2017-12-31"
    level_events = [
        {"name": "L1", "index": event_date, "c0": 5.0},
    ]
    bias = 4.0
    obs_scale = 0.5
    ds = pd.date_range(start="2016-01-31", end="2021-05-31", freq="M")
    T = len(ds)
    y = bias + np.random.normal(scale=obs_scale, size=T)
    for level_i in level_events:
        delta_i = ds >= event_date
        y += delta_i * level_i["c0"]
    return {
        "data": pd.Series(y, index=ds),
        "level_events": level_events,
        "trend_events": [],
    }


def test_level_event(create_single_event_data):
    jp = JudgyProphet()
    jp.fit(**create_single_event_data)
    model_output = jp.get_model_output()
    assert np.isclose(model_output["level_events"][0]["c"], 5.0, atol=1e-1)
    all_unspecified_params = [
        param
        for cp in model_output["changepoint_events"]
        for param in [cp["cp_c"], cp["cp_m"]]
    ]
    assert np.allclose(all_unspecified_params, 0, atol=1e-2)
    assert np.isclose(model_output["base"]["beta_0"], 4.0, atol=1.0)
    predictions = jp.predict(horizon=5)
    # Fetch expected
    ds = pd.date_range(start="2016-01-31", end="2021-10-31", freq="M")
    delta_i = ds >= "2017-12-31"
    expected = 4.0 + delta_i * 5.0
    assert np.allclose(predictions["forecast"], expected, atol=1.0)


judgyprophet/utils.py
import logging
import warnings
from typing import Type

logger = logging.getLogger(__name__)


def log_and_raise(msg: str, error: Type[BaseException]):
    """
    Add an error to logging then raise the corresponding error.

    :param: msg (string) the message to print to log.
    :param: error the error to raise.
    ...
    :raises: error of type `error` param.
    """
    logger.error(msg)
    raise error(msg)


def log_and_warn(msg: str):
    """
    Add an error to logging then raise the corresponding error.

    :param: msg (string) the message to print to log.
    :param: error the error to raise.
    ...
    :raises: error of type `error` param.
    """
    logger.warning(msg)
    warnings.warn(msg)


def assert_log_raise(condition: bool, msg: str, error: Type[BaseException]):
    """
    If condition is False, log then raise corresponding error.

    :param: condition (bool) condition to check true.
    :param: msg (string) the message to print to log.
    :param: error the error to raise.
    ...
    :raises: error of type `error` param.
    """
    if not condition:
        logger.error(msg)
        raise error(msg)


def assert_log_warn(condition: bool, msg: str):
    """
    If condition is False, log then raise warning.

    :param: condition (bool) condition to check true.
    :param: msg (string) the message to print to log.
    ...
    :raises: warning if condition false.
    """
    if not condition:
        logger.warning(msg)


judgyprophet/tutorials/resources.py
from typing import Any, Dict, List

import numpy as np
import pandas as pd
import seaborn as sns
from prophet import Prophet


def get_level_event() -> pd.Series:
    """Get example data for a level event tutorial"""
    dates = pd.date_range(start="2019-01-01", end="2021-01-01", freq="MS")
    T = len(dates)
    level_time = "2020-03-01"
    level_amount = 8

    # Set temporary seed for reproducibility
    state = np.random.get_state()
    np.random.seed(13)
    observations = np.random.normal(loc=4, size=T)
    np.random.set_state(state)
    observations += (dates > level_time) * level_amount
    output = pd.Series(observations, index=dates)
    return output


def get_trend_event(uptake: bool = True) -> pd.Series:
    """Get example data for a trend event tutorial"""
    dates = pd.date_range(start="2019-06-01", end="2021-06-01", freq="MS")
    T = len(dates)
    trend_time = "2020-09-01"
    base_trend = 1
    trend_coefficient = 4

    # Set temporary seed for reproducibility
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(loc=4, size=T)
    np.random.set_state(state)
    if uptake:
        observations += np.cumsum(dates > trend_time) * trend_coefficient
    else:
        observations -= np.cumsum(dates > trend_time) * trend_coefficient
    output = pd.Series(observations, index=dates)
    return output


def get_damped_trend_event() -> pd.Series:
    """Get example data for a trend event tutorial"""
    dates = pd.date_range(start="2019-06-01", end="2021-06-01", freq="MS")
    T = len(dates)
    trend_time = "2020-01-01"
    base_trend = 0.5
    trend_coefficient = 4
    damping = 0.8

    # Set temporary seed for reproducibility
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(loc=4, size=T)
    np.random.set_state(state)
    delta = dates > trend_time
    t = np.cumsum(delta)
    observations += delta * trend_coefficient * (1 - damping**t) / (1 - damping)
    output = pd.Series(observations, index=dates)
    return output


def get_additive_seasonality_linear_trend() -> pd.Series:
    """Get example data for additive seasonality tutorial"""
    dates = pd.date_range(start="2017-06-01", end="2021-06-01", freq="MS")
    T = len(dates)
    base_trend = 2
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(loc=4, size=T)
    np.random.set_state(state)
    seasonality = 12
    time = np.arange(0, T / seasonality, 1 / seasonality)
    amplitude = 10
    sin_cos_wave = amplitude * np.cos(2 * np.pi * time) + amplitude * np.sin(
        2 * np.pi * time
    )
    observations += sin_cos_wave
    output = pd.Series(observations, index=dates)
    return output


def get_multiplicative_seasonality_linear_trend() -> pd.Series:
    """Get example data for multiplicative seasonality tutorial"""
    dates = pd.date_range(start="2017-06-01", end="2021-06-01", freq="MS")
    T = len(dates)
    base_trend = 2
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(loc=4, size=T)
    np.random.set_state(state)
    seasonality = 12
    time = np.arange(0, T / seasonality, 1 / seasonality)
    amplitude = 6
    theta = 10
    sin_wave = amplitude * np.sin(2 * np.pi * time) + theta
    observations *= sin_wave
    output = pd.Series(observations, index=dates)
    return output


def get_additive_seasonal_damped_trend_event() -> pd.Series:
    """Get example data for additive seasonality and a trend event tutorial"""
    dates = pd.date_range(start="2015-01-01", end="2021-12-01", freq="MS")
    T = len(dates)
    trend_time = "2018-01-01"
    base_trend = 0.2
    trend_coefficient = 9
    damping = 0.9

    # Set temporary seed for reproducibility
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(scale=5, loc=10, size=T)
    np.random.set_state(state)
    delta = dates > trend_time
    t = np.cumsum(delta)
    observations += delta * trend_coefficient * (1 - damping**t) / (1 - damping)

    # add sinus function for periodicity
    seasonality = 6
    time = np.arange(0, T / seasonality, 1 / seasonality)
    theta = 1.5
    amplitude = 12
    sinewave_1 = amplitude * np.sin(2 * np.pi * 0.5 * time + theta)
    sinewave_2 = amplitude / 2 * np.sin(2 * np.pi * 1 * time + theta)
    sinewave_3 = amplitude / 3 * np.sin(2 * np.pi * 1.5 * time + theta)
    observations += sinewave_1
    observations += sinewave_2
    observations += sinewave_3
    output = pd.Series(observations, index=dates)
    return output


def get_additive_seasonal_damped_trend_event_correlated_noise() -> pd.Series:
    """Get example data for additive seasonality and a trend event tutorial"""
    dates = pd.date_range(start="2015-01-01", end="2021-12-01", freq="MS")
    T = len(dates)
    trend_time = "2018-01-01"
    base_trend = 0.2
    trend_coefficient = 9
    damping = 0.9

    # Set temporary seed for reproducibility
    state = np.random.get_state()
    np.random.seed(13)
    observations = base_trend * np.arange(T) + np.random.normal(scale=5, loc=10, size=T)
    np.random.set_state(state)
    delta = dates > trend_time
    t = np.cumsum(delta)
    observations += delta * trend_coefficient * (1 - damping**t) / (1 - damping)

    # add correlated noise
    noise_start = "2018-04-01"
    noise_end = "2018-08-01"
    noise_trend_coefficient = -8
    noise_delta = (dates >= noise_start) & (dates <= noise_end)
    t = np.cumsum(noise_delta)
    observations += (
        noise_delta * noise_trend_coefficient * (1 - damping**t) / (1 - damping)
    )

    # add sinus function for periodicity
    seasonality = 6
    time = np.arange(0, T / seasonality, 1 / seasonality)
    theta = 1.5
    amplitude = 12
    sinewave_1 = amplitude * np.sin(2 * np.pi * 0.5 * time + theta)
    sinewave_2 = amplitude / 2 * np.sin(2 * np.pi * 1 * time + theta)
    sinewave_3 = amplitude / 3 * np.sin(2 * np.pi * 1.5 * time + theta)
    observations += sinewave_1
    observations += sinewave_2
    observations += sinewave_3
    output = pd.Series(observations, index=dates)
    return output


def base_prophet_forecast(
    actuals: pd.Series, cutoff: str, events: List[Dict[str, Any]]
) -> pd.DataFrame:
    """Make a base forecast using prophet"""
    changepoint_list = [
        event_dict["index"]
        for event_dict in events
        if pd.to_datetime(event_dict["index"]) <= pd.to_datetime(cutoff)
    ]
    prophet_model = Prophet(changepoints=changepoint_list)
    cutoff_actuals = (
        actuals[:cutoff].reset_index().rename(columns={"index": "ds", 0: "y"})
    )
    prophet_model.fit(cutoff_actuals)
    future_df = actuals.reset_index().rename(columns={"index": "ds", 0: "y"})
    future_df = future_df[["ds"]]
    predictions = prophet_model.predict(future_df)
    predictions = predictions[["ds", "yhat"]].rename(columns={"yhat": "value"})
    predictions["insample"] = pd.to_datetime(predictions["ds"]) <= pd.to_datetime(
        cutoff
    )
    predictions["method"] = "prophet"
    return predictions


def plot_forecast(
    actuals: pd.Series,
    predictions: pd.Series,
    cutoff: str,
    events: List[Dict[str, Any]],
):
    """Plot a forecast that's cutoff before actuals finish"""
    base_prophet_df = base_prophet_forecast(
        actuals=actuals, cutoff=cutoff, events=events
    )
    # Find cutoff date for plot
    plot_cutoff = min(
        [base_prophet_df.ds.max(), actuals.index.max(), predictions.index.max()]
    )
    predict_df = (
        predictions.reset_index()
        .rename(columns={"index": "ds", "forecast": "value"})
        .assign(method="judgyprophet")
        .loc[:, ["ds", "value", "insample", "method"]]
    )
    actuals_df = (
        actuals.loc[:cutoff]
        .reset_index()
        .rename(columns={"index": "ds", 0: "value"})
        .assign(method="actuals", insample=True)
    )
    future_actuals_df = (
        actuals.loc[cutoff:]
        .reset_index()
        .rename(columns={"index": "ds", 0: "value"})
        .assign(method="actuals", insample=False)
    )
    plot_df = (
        pd.concat([predict_df, actuals_df, future_actuals_df, base_prophet_df])
        .reset_index(drop=True)
        .query(f"ds <= '{plot_cutoff}'")
    )
    sns.lineplot(
        data=plot_df,
        x="ds",
        y="value",
        hue="method",
        style="insample",
        style_order=[True, False],
    )
